### ISSUE TITLE
Narrow type of variable when declared as literal (also tuples)

### ISSUE THREAD

**Issue by Nathan-Fenner** [Microsoft/TypeScript#16896]
**TypeScript Version:**  2.4.1

**Code**

The following doesn't compile. I think it should. I'll outline how below:

```ts
function blah(arg: "foo" | "bar") {
}
let x = "foo";
blah(x);
x = "something else";
```

**Desired behavior:**

This would compile.

**Actual behavior:**

`demo.ts(4,6): error TS2345: Argument of type 'string' is not assignable to parameter of type '"foo" | "bar"'.`

**Suggestion**

The type of `x` should be inferred to be `string`, but it should be narrowed to `"foo"` via flow-sensitive typing, much like the below program:

```ts
function blah(arg: "foo" | "bar") {
}
let x = "foo";
if (x != "foo") throw "impossible"; // NOTE: codegen would NOT insert this
blah(x);
x = "something else";
```

Note this this code currently compiles without any problem: `x` is inferred to have type `string`, but the guard immediately narrows it to type `"foo"`. The assignment at the bottom widens `x`'s type back to `string`.

I would find it useful if TS behaved like this without the need for an unnecessary guard. Specifically, my proposal is that when encountering a declaration such as 

```ts
let variable1 = "some literal"; // inferred as variable1: string, like today
let variable2 = 123; // inferred as variable2: number, like today
let variable3: string = "some literal"
let variable4: number = 123;
```

(in addition to `var` and `const`)

would in each case be narrowed immediately via flow-sensitive typing to the specific literal type that they're assigned.

This suggestion doesn't affect assignment (although it could, I would find it much less immediately helpful). It also doesn't affect assignment to anything more complicated than a literal. Objects containing literals, function calls, or `as` casts would not have any different behavior than they do today.

** Syntax Changes **

Nothing changes in syntax.

** Code Generation **

Nothing changes in code generation.

** Reverse Compatibility **

Due to `x` being assigned a more-precise type than it was in the past, some code is now marked as unreachable that previously:

```ts
let x = "foo";
if (x == "bar") { // error TS2365: Operator '==' cannot be applied to types '"foo"' and '"blah"'.
  // ...
}
```

I don't know how commonly this occurs in production code for it to be a problem. If it is common, this change could be put behind a flag, or there could be adjustments made to these error cases so that the narrowing doesn't occur (or still narrows, but won't cause complaints from the compiler in these error cases).

If the existing behavior is strongly needed, then the change can be circumvented by using an `as` cast or an indirection through a function, although these are a little awkward:

```ts
let x = "foo" as string;
let x = (() => "foo")()
```

## Tuples

In addition, I'd like the same behavior with tuples. i.e., the following could should compile:

```ts
function blah(arg: [number, number]) {
}
let x = [1, 3];
blah(x);
x = [1, 2, 3];
```

with `x` again having inferred type `number[]` but being flow-typed to `[number, number]`. The same basic considerations apply here as above.

Note that I don't expect them to be applied "in combination" (that is, `let x = [1, 2, 3]` will have narrowed type `[number, number, number]`, not `[1, 2, 3]`, and inferred type `number[]`, as before).

See https://github.com/Microsoft/TypeScript/issues/16276 and https://github.com/Microsoft/TypeScript/issues/16360 and probably others for related but different approaches to take here.

**Comment by aj-r** [comment]
Agreed - it would be really nice to make type narrowing work for object literals. Also consider this case:

```
interface Foo {
   a: string | undefined;
}

function bar() {
    const foo: Foo = { a: "test string" };
    // Currently does not compile, because the compiler thinks a is possibly undefined:
    return foo.a.length;
}
```

It would be nice if the above code compiled (with strictNullChecks enabled) because `a` is clearly not undefined.

**Comment by LinusU** [comment]
The more I think about this the more I think it would be highly desirable. I cannot think of any case that would be negatively impacted by making literals be a more narrow type since you can always widen the type. But I can see a lot of positive changes coming from this since you cannot automatically narrow a wide type.

Consider these examples:

```typescript
declare interface Animal { type: 'cat' | 'dog' }
declare function interactWith (animal: Animal): void

const charlie = { type: 'dog' }

// Argument of { type: string } is not assignable to Animal
interactWith(charlie)
```

A fix in this simple case would be to add e.g. ` as Animal` to `const charlie = ...`.

```typescript
type Schema = NullSchema | BooleanSchema | NumberSchema | StringSchema | ArraySchema | ObjectSchema
interface NullSchema { type: 'null' }
interface BooleanSchema { type: 'boolean' }
interface NumberSchema { type: 'number' }
interface StringSchema { type: 'string' }
interface ArraySchema { type: 'array', items: Schema }
interface ObjectSchema { type: 'object', properties: Record<string, Schema> }

declare function validate (schema: Schema, input: unknown): boolean

const schema = {
  type: 'object'
  properties: {
    name: { type: 'string' }
    items: {
      type: 'array',
      items: {
        type: 'object',
        properties: {
          type: { type: 'string' }
        }
      }
    }
  }
}

// Argument of ... is not assignable to Schema
validate(schema, null)
```

Here the fix is quite annoying, you need to go in and declare the type of every nested schema by making all strings string literals.

```typescript
const schema = {
  type: 'object' as 'object'
  properties: {
    name: { type: 'string' as 'string' }
    items: {
      type: 'array' as 'array',
      items: {
        type: 'object' as 'object',
        properties: {
          type: { type: 'string' as 'string' }
        }
      }
    }
  }
}
```

I would love it if we could move forward on this. Is there something that I could contribute with? Do we need to do some kind of research on how this would affect current projects using TypeScript?

**Comment by Richiban** [comment]
@LinusU 

> Here the fix is quite annoying, you need to go in and declare the type of every nested schema by making all strings string literals.

Surely the workaround for this is to simply add a type annotation? `const schema : Schema = {`, so:

```typescript
const schema : Schema = {
  type: 'object',
  properties: {
    name: { type: 'string' },
    items: {
      type: 'array',
      items: {
        type: 'object',
        properties: {
          type: { type: 'string' }
        }
      }
    }
  }
}
```

**Comment by LinusU** [comment]
@Richiban, unfortunately, that workaround will throw away any information on what type of scheme it is, so you would no longer be able to do `schema.properties`.

The logical next step is to declare it as `const schema: ObjectSchema = ...` then, but that throws away any type information about the properties, so then you can no longer do `schema.properties.name`...

**Comment by pirix-gh** [comment]
This issue is very related to https://github.com/microsoft/TypeScript/pull/29510

### Proposal

I think that it is important to give full control on type inference. Just a part of the puzzle is missing, because most of the work was done in https://github.com/microsoft/TypeScript/pull/29510. So I propose introducing a new keyword `as self` that works like `as const` but without making the type deeply immutable.

This would also give full power to the user over generics (not having to explicitly pass them):

```ts
const f = <T>(t: T) => t

// default, widening behavior
const test0 = f({a: {b: 'hello'}})
// => {
//     a: {
//         b: string;
//     };
// }

// non-widening, immutable behavior
const test1 = f({a: {b: 'hello'}} as const)
// => {
//     readonly a: {
//         readonly b: 'hello';
//     };
// }

// non-widening, mutable behavior
const test2 = f({a: {b: 'hello'}} as self)
// => {
//     a: {
//         b: 'hello';
//     };
// }
```

This could give full control on type inference, in general. So we could either use `as const`, `as self`, or nothing (as needed).

**Comment by JCMais** [comment]
For Tuples the way I'm doing it currently is by using `as const` alongside a `Writeable` helper to remove the `readonly` modifier, something like this:

```
const items = [1, 2, 3] as const

function sum3Numbers(a: [number, number, number]) {
    const [n1, n2, n3] = a
    return n1 + n2 + n3
}

// should give an error
sum3Numbers(items)

// workaround:
type Writeable<T> = { -readonly [P in keyof T]: T[P] }
sum3Numbers(items as Writeable<typeof items>)
```
[Playground Link](http://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=13&pc=46#code/MYewdgzgLgBAllApgWwjAvDA2gRgDQwBMBAzALowCGaokUAUPQGYCuYwUc4MELyJAOT4AjRACcIACkoAubGBHiCC5KLHLFYsgEoYAb3owjMWtHn4YYYpfIYqh42MRQWYsJZwwA1JcLeb9AC+jAD0ITwAFiAsADYAJjAA5nAAbohU7uJiIGL0vPxCquJSCCgQ2qHhAO45ANaU2WxxMvRQAJ4ADukA6mKllMIxiAA8ACoAfHZ6MAC0TpRx4DFt2AAK8O61iG0gTDCjZHKjWKsUwfmCmiVIqFRovf2DI+1du-A3EOPaQA)
